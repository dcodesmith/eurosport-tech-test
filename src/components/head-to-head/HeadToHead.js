import React, { Component } from 'react';
import sortBy from 'lodash.sortby';

import Stats from '../stats/';
import Labels from '../labels/';
import LoadingMessage from '../loading-message/';
import fetchPlayers from '../../services/fetchPlayers';

class HeadToHead extends Component {
  _isMounted = false;
  state = {
    players: []
  };

	async componentDidMount() {
    this._isMounted = true;
    const { players } = await fetchPlayers();

    if (this._isMounted) {
      this.setState({ players: sortBy(players, [player => player.data.rank]) });
    }
	}

  componentWillUnmount() {
    this._isMounted = false;
  }

  render() {
    const { players } = this.state;
    const [ firstPlayer, secondPlayer ] = players;

    if (!players.length) {
      return <LoadingMessage />;
    }

    return (
      <div className="grid">
        <Stats player={firstPlayer} />
        <Labels />
        <Stats player={secondPlayer} />
      </div>
    )
  }
}

export default HeadToHead;

