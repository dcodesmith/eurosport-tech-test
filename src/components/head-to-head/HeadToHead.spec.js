import React from 'react';
import { shallow } from 'enzyme';
import mockfetchPlayers from '../../services/fetchPlayers';

import HeadToHead from './';
import mockPlayers from '../../fixtures/players';
jest.mock('../../services/fetchPlayers');

describe('Given a HeadToHead Component', () => {
	describe('When rendered', () => {
    let HeadToHeadComponent;

    describe('and players have not been fetched', () => {
      beforeAll(() => {
        HeadToHeadComponent = shallow(<HeadToHead />);
      });

      it('should render the Loading Message component', () => {
        expect(HeadToHeadComponent.find('LoadingMessage')).toMatchSnapshot();
      });
    });

    describe('and players have been fetched', () => {
      beforeAll(() => {
        mockfetchPlayers.mockImplementationOnce(() => 
          Promise.resolve({ players: mockPlayers })
        );

        HeadToHeadComponent = shallow(<HeadToHead />);
      });

      it('should not render the Loading Message component', () => {
        expect(HeadToHeadComponent.find('LoadingMessage')).toMatchSnapshot();
      });

      it('should render a Label component', () => {
        expect(HeadToHeadComponent.find('Labels')).toMatchSnapshot();        
      });

      it('should render a Stats component for each players with the correct props', () => {
        const statsComponents = HeadToHeadComponent.find('Stats');
        const firstPlayerStatsComponent = statsComponents.at(0);
        const secondPlayerStatsComponent = statsComponents.at(1);

        expect(HeadToHeadComponent.find('Stats').length).toEqual(2);

        expect(firstPlayerStatsComponent.prop('player')).toEqual(mockPlayers[1]);
        expect(secondPlayerStatsComponent.prop('player')).toEqual(mockPlayers[0]);
      });
    });
	});
});