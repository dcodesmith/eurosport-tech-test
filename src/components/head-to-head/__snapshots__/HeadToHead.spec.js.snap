// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a HeadToHead Component When rendered and players have been fetched should not render the Loading Message component 1`] = `null`;

exports[`Given a HeadToHead Component When rendered and players have been fetched should render a Label component 1`] = `<Labels />`;

exports[`Given a HeadToHead Component When rendered and players have not been fetched should render the Loading Message component 1`] = `<LoadingMessage />`;
